### DISCUSSION TITLE
Display a dialog when an unhandled exception occurs

### DISCUSSION THREAD

**Discussion by flarive** [AvaloniaUI/Avalonia#14200] [Q&A]
Hello,

I would like to be able to display a message box (dialog) when any unhandled exception occur in my Avalonia app.

I have discovered that in Program.cs it's too late to display anything because the app is being killed :

 ```
public static void Main(string[] args)
 {
     try
     {
         BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);
     }
     catch (Exception ex)
     {
           // too late !
           // can't display any dialog from here
     }
 }
```

So I would like to add a global exception catcher in App.xaml.cs to catch unhandled exceptions before they are catched by Program.cs.
But registering to AppDomain.UnhandledException don't work, i don't know how to catch exceptions in App.xaml.cs :(

For info, i dont use ReactiveUI and i don't want to add it to my project :)

Thanks for your help !

**Comment by timunie** [comment]
@flarive probably better to run a new exe showing the error message.

**Reply by flarive** [reply]
Hum, i don't like this solution :(
It should be possible to do better no ?

I just want to be able to catch exceptions somewhere before Program.cs to be able to manage it on my side.

**Reply by thevortexcloud** [reply]
Ideally you should be just catching the exceptions before they even get to there. It's somewhat of a code smell if you need to do this as far as I am concerned.

**Reply by timunie** [reply]
You cannot know if Avalonia is in a corrupted state here, so I doubt there is a better place to do this. If you knew the exception in advance, you wouldn't hit that place. 

related: #7988

**Reply by StefanKoell** [reply]
+1 for me as well. An unhandled exception should also not always terminate the app. Larger applications (with plugins, for example) shouldn't always die when an unhandled exception is thrown. It should be logged/reported and the user should have the option to continue. I know there might be a good chance that the application is ending up in a broken state which required a restart but in my apps I have right now, it is a viable option and 95% of the unhandled exceptions allow you to continue to use the app.

**Reply by flarive** [reply]
I have tried to catch exceptions in App.axaml.cs but nothing is catched.
I'm using AppDomain.UnhandledException event...
Do you know another way to catch unhandled exceptions outside Program.cs ?

**Reply by thevortexcloud** [reply]
The docs say you should catch your main method.

https://docs.avaloniaui.net/docs/concepts/unhandledexceptions

**Reply by StefanKoell** [reply]
@thevortexcloud this would always terminate the app, correct? Is there a way to keep the app running?

**Reply by thevortexcloud** [reply]
I am not sure, I have never tried it.

**Reply by timunie** [reply]
not implemented rn. There were some attempts earlier, but we had no luck yet to provide a stable solution. SO for now, one should handle the exceptions as early as possible. Probably a helper class can do.  

- https://github.com/AvaloniaUI/Avalonia/issues/8418
- https://github.com/AvaloniaUI/Avalonia/pull/13476 (this gives some hope though)

**Reply by StefanKoell** [reply]
Got it, thanks. Maybe this can be improved and provided like on WinForms, WPF, etc.

I think it's an important "feature" when you have to deal with code and components where you don't have access to the source. An unhandled exception of a 3rd party component should never tear down the whole application.

**Reply by thevortexcloud** [reply]
> unhandled exception of a 3rd party component should never tear down the whole application.

I have written plugins for several programs where this does happen. It just forces people to write better code or catch the exceptions themselves. If someone is writing bad code it's not really your problem at that point.

**Reply by StefanKoell** [reply]
Agreed, in an ideal world. Unfortunately it's not always in your hand and the end user is complaining about the "app" not working and goes to the competition. My app uses a lot of 3rd party code (RDP, Terminal, Browser, etc.). There are plenty unhandled exceptions which are annoying but if a user has 25 tabs open and one tab crashes for whatever reason, you don't want to bring down the whole app (think of a web browser). In .NET this shouldn't be a big problem and in WPF and WinForms it's easy to handle.

**Reply by flarive** [reply]
I'm now able to catch exceptions by using AppDomain.CurrentDomain.UnhandledException in Program.cs only if remove the try/catch/finally arround BuildAvaloniaApp().StartWithClassicDesktopLifetime(args);

But it doesn't prevent the app from closing because of that exception :(

**Reply by jp2masa** [reply]
> There are plenty unhandled exceptions which are annoying but if a user has 25 tabs open and one tab crashes for whatever reason, you don't want to bring down the whole app (think of a web browser).

This isn't really a good example. Most web browsers now run tabs as separate processes. In your specific case, why not add `try..catch` around plugin calls?

**Reply by StefanKoell** [reply]
@jp2masa I wasn't talking strictly technical. The browser example was just a way to explain the pain when one tab crashes and the whole app dies and work is lost. I know that browsers work differently. Your suggestion to just wrap all the calls in a `try...catch` is also not as simple as you might think.

Let me explain the scenario:
* a bigger/more complex app (common in enterprises or commercial apps) allows you to load 1st and 3rd party plugins
* plugins can provide their own UI (let's say the content in tabs in a tabbed application)
* these plugins can execute code on their own (imagine a user control with a button/command which executes code)
* there's no way I can wrap any `try...catch` around that code because I don't see the code and I don't control the execution of the code
* the argument to just write better/stable code to tell the plugin developer is not really helping. It's basically the same as demanding from every developer to write bug free code - which is just ridiculous. 

In my case it is about a commercial application where 3rd party developers can provide plugins. When a plugin causes the whole app to crash it doesn't really help me to paint the blame to the plugin author. The customer is upset and switches product.

Avalonia is .NET based which allows dynamic loading of assemblies/plugins. Other technologies (like WinForms) can deal with that scenario. I know because my app is currently a WinForms app and this is how I do it now. If every unhandled exception from one of the plugins or component vendors would take down the whole app, I wouldn't have a business anymore.

So this is, in my opinion, a critical feature for these kinds of applications to be successful with Avalonia.

What I really would like to know from the core team or maintainers (@MikeCodesDotNET) is, if this is addressed at some point or worse: if there's a road block which makes it impossible to implement. If it's the latter, I'm really f*#$ed, I guess.

**Reply by timunie** [reply]
@StefanKoell have you checked if https://github.com/AvaloniaUI/Avalonia/pull/13476 may help you?

**Reply by StefanKoell** [reply]
@timunie I don't know, maybe? If this PR allows me to catch all the unhandled exceptions, then yes. Not sure if it covers that all.

**Reply by StefanKoell** [reply]
@timunie I just tried the linked PR but I failed to test it. I'm unable to create the packages based on that PR. As I can see, this is still not merged and available in the nightly builds. What can I do to test this? Why isn't it merged yet. Can we get it into the nightly build as an "opt-in" feature so that setting a feature-flag or something enables this "feature"?
